### ISSUE TITLE
Embed DNS into kube-proxy?

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#7470]
DNS is basically non-optional at this point, and this is more fault-tolerant and performant.  Consider.

**Comment by bakins** [comment]
I'd rather keep it a separate component. Separation of concerns and such.  Perhaps a DNS component that we run on every node rather than the current service based approach?

**Comment by quinton-hoole** [comment]
@thockin I can understand and agree that something like this is worthy of consideration, but I think that we need to be able to make services (of which DNS is currently one) fault-tolerant and performant, otherwise we have bigger problems.  I share @bakin's concerns about bundling it into kube-proxy (which I hope shrinks to become a relatively simple iptables nanny when we move the proxy logic out of it into iptables).

@bakins What is your concern regarding the current service based approach?  If we run the DNS servers on each node as some sort of native component, we end up having to solve all of the challenges surrounding deployment, upgrade, monitoring etc all over again, surely?  Those are the things that services are already designed to do well.

**Comment by bakins** [comment]
@quinton-hoole separate component for DNS was quick thought as a "middle-ground" between current implementation and building it into kube-proxy.  Your comments are spot-on WRT deployment, etc.  Catching up on the other conversations concerning DNS while I think.

**Comment by errordeveloper** [comment]
Looks like this issue had been open for a while, however I thought I should mention that some of the design decisions in `weavedns` could be of interest.

  - [WeaveDNS Requirements
](https://docs.google.com/document/d/1WkKyNd1NwfRDlEr_z9UCAYCS6_ZXOPQVhWEuFiQJek0/edit?usp=sharing)
  - [WeaveDNS Design Alternatives](https://docs.google.com/document/d/1RClf6zvuoNSLDxKWta-Xv8G6eYuTF-NhfSQgLFWxMIw/edit?usp=sharing)
  - [Implementation](https://github.com/weaveworks/weave/tree/master/nameserver)

**Comment by thockin** [comment]
There's an open PR, too, but not enough time to review everything :(  I
hope to circle back to this topic before 1.1, but we'll see.  Thanks for
extra info.

On Fri, Aug 21, 2015 at 7:57 AM, Ilya Dmitrichenko <notifications@github.com

**Comment by errordeveloper** [comment]
For the record, @tomwilkie's [blog post](http://blog.weave.works/2015/09/08/weave-gossip-dns/) describes it at pretty high level, as we have just released it in Weave 1.1 today.

**Comment by thockin** [comment]
I'd be happy to consider it as an alternative to our existing DNS if we can
make it work and it performs better.  I tried to get dnsperf to compile on
Debian but failed after 20 minutes of trying to get the necessary deps
installed.

**Comment by tomwilkie** [comment]
@bboreham did a benchmark for his http://softwarecircus.eu/ talk tomorrow:

http://go-talks.appspot.com/github.com/bboreham/talks/camp.slide#10

Looks like weavendns is about 30% faster than skydns.  It wasn't a massively scientific test apparently.

**Comment by thockin** [comment]
If you guys want to pony up some time, I'm happy to guide - I have no
religious affiliation to any particular DNS solution.  One thing worth
noting is that SkyDNS has the ability to provide backends that are not
etcd, so we can slurp data straight from our API (once we get around to
using this :)

**Comment by thockin** [comment]
Now that we have a standalone DNS server and a cache in front of it, I'd like to evaluate (via bug reports and perf complaints) the need for this.

**Comment by rtreffer** [comment]
For DNS we'd like the following things to be considered

1. Set multiple DNS servers via kubelet
2. Setup a DNS server/cache per host

**kubelet can inject only 1 DNS server** as the cluster-dns option is parsed into net.IP. (docker supports specifying --dns multiple times)
Why this is important: Having only one server will cause 5s timeouts, while two servers will cause glibc to at least try the second server as well and musl will go ahead and ask both server in parallel, thus improving latencies.
A per-host DNS cache is highly recommended as docker images usually do not have a DNS cache. Thus all queries go over the network and a single packet drop will cause a 5s timeout.
See http://man7.org/linux/man-pages/man5/resolv.conf.5.html and
http://wiki.musl-libc.org/wiki/Functional_differences_from_glibc#Name_Resolver_.2F_DNS

We are currently **moving to a DNS daemonset**, too. However that has a problem with docker-proxy.
The docker-proxy runs on all interfaces (--host 0.0.0.0). If we inject the host ip as cluster dns a pod container to query <hostip>:53 for DNS. docker-proxy forwards that correctly but once the response arrives at docker-proxy it will simply send a packet back. As the socket is listening on 0.0.0.0 the best interface for responding is picked, which is the bridge. Thus the response is coming from <bridgeip>:53 which gets dropped by the pod. (I'm not sure if docker-proxy should listen on 0.0.0.0 for a hostport)

Using host network works, but I'm still unhappy with the only-one-dns-server situation as this will cause unnecessary breakage in the case of a permanent name server failure on a node.

Ideally I'd like to inject 2 DNS servers, an on-host DNS server as a daemonset and a service that is backed by the same daemonset.

We heavily rely on DNS and already see >30k dns queries / second inside our cluster.

**Comment by thockin** [comment]
On Sep 22, 2016 7:04 AM, "Rene Treffer" <notifications@github.com> wrote:
>
> For DNS we'd like the following things to be considered
>
> Set multiple DNS servers via kubelet
> Setup a DNS server/cache per host
>
> kubelet can inject only 1 DNS server as the cluster-dns option is parsed
into net.IP. (docker supports specifying --dns multiple times)
> Why this is important: Having only one server will cause 5s timeouts,
while two servers will cause glibc to at least try the second server as
well and musl will go ahead and ask both server in parallel, thus improving
latencies.

Having two 'nameserver' entries will double the DNS load on the cluster.
We are actively investigating how to make DNS more reliable and
performance.  I considered this route, but I don't know if it is the best
approach.

> A per-host DNS cache is highly recommended as docker images usually do
not have a DNS cache. Thus all queries go over the network and a single
packet drop will cause a 5s timeout.

We are looking at this as an option for near term.

> See http://man7.org/linux/man-pages/man5/resolv.conf.5.html and
> http://wiki.musl-libc.org/wiki/Functional_differences_
from_glibc#Name_Resolver_.2F_DNS
>
> We are currently moving to a DNS daemonset, too. However that has a
problem with docker-proxy.
> The docker-proxy runs on all interfaces (--host 0.0.0.0). If we inject
the host ip as cluster dns a pod container to query :53 for DNS.
docker-proxy forwards that correctly but once the response arrives at
docker-proxy it will simply send a packet back. As the socket is listening
on 0.0.0.0 the best interface for responding is picked, which is the
bridge. Thus the response is coming from :53 which gets dropped by the pod.
(I'm not sure if docker-proxy should listen on 0.0.0.0 for a hostport)

I will have to look into this.  I don't know docker-proxy well, but that
seems like a pretty broken problem.  At least in upstream kube, you should
not need docker-proxy any more.

> Using host network works, but I'm still unhappy with the
only-one-dns-server situation as this will cause unnecessary breakage in
the case of a permanent name server failure on a node.
>
> Ideally I'd like to inject 2 DNS servers, an on-host DNS server as a
daemonset and a service that is backed by the same daemonset.
>
> We heavily rely on DNS and already see >30k dns queries / second inside
our cluster.

**Comment by kjvalencik** [comment]
@rtreffer We also moved cluster DNS to a daemonset. To help reduce the scope of failure, we kept kubedns as an RC in the cluster.

1. Created the `kube-dns` RC. No changes.

2. Changed `kube-dns` `clusterIP` from `10.100.0.10` to `10.100.0.11`:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: kube-dns
  namespace: kube-system
  labels:
    k8s-app: kube-dns
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: KubeDNS
spec:
  clusterIP: 10.100.0.11
  selector:
    k8s-app: kube-dns
  ports:
  - name: dns
    port: 53
    protocol: UDP
    targetPort: 53
  - name: dns-tcp
    port: 53
    protocol: TCP
    targetPort: 53
```

3. Added a `DaemonSet` with `dnsmasq` pointed back at `kube-dns`. This uses `hostNetwork` but, could also use a node port. The advantage to `hostNetwork` is that you can limit it to listening on `localhost`.
```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: dnsmasq
  namespace: kube-system
  labels:
    app: dnsmasq
spec:
  selector:
    matchLabels:
      app: dnsmasq
  template:
    metadata:
      labels:
        app: dnsmasq
    spec:
      hostNetwork: true
      containers:
        - name: dnsmasq
          image: gcr.io/google_containers/kube-dnsmasq-amd64:1.4
          imagePullPolicy: IfNotPresent
          args:
            - --cache-size=1000
            - --no-resolv
            - --server=10.100.0.11#53
            - --listen-address=127.0.0.1
            - --log-facility=-
          livenessProbe:
            failureThreshold: 5
            httpGet:
              path: /healthz
              port: 8080
              scheme: HTTP
            initialDelaySeconds: 60
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 5
          ports:
            - containerPort: 53
              name: dns
              protocol: UDP
            - containerPort: 53
              name: dns-tcp
              protocol: TCP
        - name: healthz
          image: gcr.io/google_containers/exechealthz-amd64:1.2
          imagePullPolicy: IfNotPresent
          args:
            - --cmd=nslookup kubernetes.default.svc.cluster.local 127.0.0.1 >/dev/null
            - --url=/healthz
            - --port=8080
            - --quiet
          ports:
            - containerPort: 8080
              protocol: TCP
```

4. Add an `iptables` rule to map the cluster DNS IP to the node.
```bash
sudo iptables -t nat -I PREROUTING --dest 10.100.0.10 -p udp --dport 53 -j DNAT --to-dest 127.0.0.1
```

**Comment by frank-weindel** [comment]
@kjvalencik Thanks for this! I set this up in our production cluster with a modest `--min-cache-ttl` and `--neg-ttl`, and it solved a MAJOR performance issue we were having with DNS.

We run many pod replicas of phantomjs in our cluster, each of which is always loading webpages along with their assets (images, css, javascript, data from APIs, etc). We noticed that pretty frequently (even when not under much load) random web assets would just take 5 seconds longer to load than usual. It was always 5 seconds when we saw the issue, matching up with what @rtreffer identified as being caused by single DNS packet drops. After digging deeper we found that almost every request for a web asset was causing a DNS request to go over the wire to a RANDOM node that kube-dns was installed on (via its service endpoint). We didn't even notice this for assets being loaded from within the cluster (which is few in comparison), we noticed it with assets with external domain names.

To do implement this we had to set up scripts to run that iptables rule on boot. It would be great if there was a way to do this same thing natively within kubernetes.
